File: 110-balanced-binary-tree/balanced-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
    if (root === null) return true

    const stack = []
    const heights = new Map()
    let lastVisited = null

    let current = root

    while (stack.length > 0 || current !== null) {
        if (current !== null) {
            stack.push(current)
            current = current.left
        } else {
            let peekNode = stack[stack.length - 1]
            if (peekNode.right !== null && lastVisited !== peekNode.right) {
                current = peekNode.right
            } else {
                stack.pop()

                let leftHeight = heights.get(peekNode.left) || 0
                let rightHeight = heights.get(peekNode.right) || 0

                if (Math.abs(leftHeight - rightHeight) > 1) {
                    return false
                }

                heights.set(peekNode, 1 + Math.max(leftHeight, rightHeight))

                lastVisited = peekNode
            }
        }
    }
    return true
};

File: 110-balanced-binary-tree/balanced-binary-tree.py
# Definition for a binary tree node.
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
# class TreeNode:
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        balanced, _ = self.checkBalanced(root)
        return balanced        

    def checkBalanced(self, node: Optional[TreeNode]) -> Tuple[bool, int]:
        if not node:
            return True, 0

        left_balanced, left_height = self.checkBalanced(node.left)
        if not left_balanced:
            return False, 0

        right_balanced, right_height = self.checkBalanced(node.right)
        if not right_balanced:
            return False, 0

        current_balanced = abs(left_height - right_height) <= 1
        current_height = 1 + max(left_height, right_height)

        return current_balanced, current_height
